File: frontend/src/lib/api-client.ts
import { getToken } from './auth-token';
import { ApiResponse } from '../types/api';
import { API_BASE_URL } from './config';

// APIクライアントの設定
interface ApiClientConfig {
  baseURL?: string;
  headers?: Record<string, string>;
  timeout?: number;
}

// HTTPメソッドの型
type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';

// リクエストオプション
interface RequestOptions {
  body?: any;
  headers?: Record<string, string>;
  params?: Record<string, string>;
}

class ApiClient {
  private baseURL: string;
  private defaultHeaders: Record<string, string>;
  private timeout: number;

  constructor(config: ApiClientConfig = {}) {
    this.baseURL = config.baseURL || API_BASE_URL;
    this.defaultHeaders = config.headers || {};
    this.timeout = config.timeout || 30000;
  }

  private async getAuthHeaders(): Promise<Record<string, string>> {
    const token = getToken();
    if (token) {
      return {
        'Authorization': `Bearer ${token}`,
      };
    }
    return {};
  }

  private async request<T>(
    method: HttpMethod,
    endpoint: string,
    options: RequestOptions = {}
  ): Promise<ApiResponse<T>> {
    try {
      const authHeaders = await this.getAuthHeaders();
      const url = new URL(endpoint, this.baseURL);
      
      // URLパラメータの追加
      if (options.params) {
        Object.entries(options.params).forEach(([key, value]) => {
          url.searchParams.append(key, value);
        });
      }

      const config: RequestInit = {
        method,
        headers: {
          'Content-Type': 'application/json',
          ...this.defaultHeaders,
          ...authHeaders,
          ...options.headers,
        },
        signal: AbortSignal.timeout(this.timeout),
      };

      // ボディの追加（GET以外）
      if (options.body && method !== 'GET') {
        config.body = JSON.stringify(options.body);
      }

      const response = await fetch(url.toString(), config);
      const result = await response.json();

      // バックエンドが既にApiResponse形式で返している場合
      if ('success' in result && typeof result.success === 'boolean') {
        return result as ApiResponse<T>;
      }

      // それ以外の場合はHTTPステータスに基づいて判断
      if (!response.ok) {
        return {
          success: false,
          error: result.message || result.error || `HTTP ${response.status}: ${response.statusText}`,
        };
      }

      return {
        success: true,
        data: result,
      };
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          return {
            success: false,
            error: 'リクエストがタイムアウトしました',
          };
        }
        return {
          success: false,
          error: error.message,
        };
      }
      return {
        success: false,
        error: '予期しないエラーが発生しました',
      };
    }
  }

  // 公開メソッド
  async get<T>(endpoint: string, options?: RequestOptions): Promise<ApiResponse<T>> {
    return this.request<T>('GET', endpoint, options);
  }

  async post<T>(endpoint: string, body?: any, options?: RequestOptions): Promise<ApiResponse<T>> {
    return this.request<T>('POST', endpoint, { ...options, body });
  }

  async put<T>(endpoint: string, body?: any, options?: RequestOptions): Promise<ApiResponse<T>> {
    return this.request<T>('PUT', endpoint, { ...options, body });
  }

  async delete<T>(endpoint: string, options?: RequestOptions): Promise<ApiResponse<T>> {
    return this.request<T>('DELETE', endpoint, options);
  }

  async patch<T>(endpoint: string, body?: any, options?: RequestOptions): Promise<ApiResponse<T>> {
    return this.request<T>('PATCH', endpoint, { ...options, body });
  }
}

// デフォルトのAPIクライアントインスタンス
export const apiClient = new ApiClient();

// カスタムAPIクライアントの作成
export const createApiClient = (config: ApiClientConfig) => new ApiClient(config);

File: ios/OnlineCalendar/Modules/Common/Services/APIClient.swift
import Foundation

// MARK: - Generic HTTP Client
class APIClient {
    static let shared = APIClient()
    
    let baseURL = "http://localhost:3001/api/v1"
    private let session = URLSession.shared
    
    private init() {}
    
    // MARK: - Generic Request Methods
    func get<T: Decodable>(_ endpoint: String, type: T.Type) async throws -> T {
        return try await request(endpoint: endpoint, method: "GET", body: nil as Data?, type: type)
    }
    
    func post<T: Decodable, U: Encodable>(_ endpoint: String, body: U, type: T.Type) async throws -> T {
        let data = try JSONEncoder().encode(body)
        return try await request(endpoint: endpoint, method: "POST", body: data, type: type)
    }
    
    func put<T: Decodable, U: Encodable>(_ endpoint: String, body: U, type: T.Type) async throws -> T {
        let data = try JSONEncoder().encode(body)
        return try await request(endpoint: endpoint, method: "PUT", body: data, type: type)
    }
    
    func delete(_ endpoint: String) async throws {
        let _: EmptyResponse = try await request(endpoint: endpoint, method: "DELETE", body: nil as Data?, type: EmptyResponse.self)
    }
    
    // MARK: - Private Request Method
    private func request<T: Decodable>(
        endpoint: String,
        method: String,
        body: Data? = nil,
        type: T.Type
    ) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            print("❌ [APIClient] Invalid URL: \(baseURL)\(endpoint)")
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        // Add auth token if available
        if let token = await getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            print("📡 [APIClient] Using auth token")
        }
        
        // Add body if provided
        if let body = body {
            request.httpBody = body
        }
        
        print("🔄 [APIClient] Starting \(method) request to \(endpoint)")
        print("📡 [APIClient] \(method) \(url)")
        print("📡 [APIClient] Headers: \(request.allHTTPHeaderFields ?? [:])")
        
        do {
            let (data, response) = try await session.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                print("❌ [APIClient] Invalid response type")
                throw APIError.invalidResponse
            }
            
            print("✅ [APIClient] Response status: \(httpResponse.statusCode)")
            
            if let responseString = String(data: data, encoding: .utf8) {
                print("📡 [APIClient] Response body: \(responseString)")
            }
            
            // Handle HTTP errors
            switch httpResponse.statusCode {
            case 200..<300:
                break // Success
            case 401:
                throw APIError.unauthorized
            case 404:
                throw APIError.notFound
            case 500..<600:
                throw APIError.serverError(httpResponse.statusCode)
            default:
                throw APIError.httpError(httpResponse.statusCode)
            }
            
            // Decode response
            do {
                let decoder = createJSONDecoder()
                return try decoder.decode(T.self, from: data)
            } catch {
                print("❌ [APIClient] Decoding error: \(error)")
                throw APIError.decodingError
            }
        } catch {
            print("❌ [APIClient] Network error: \(error)")
            throw error
        }
    }
    
    // MARK: - Helper Methods
    @MainActor
    private func getAuthToken() -> String? {
        return AuthState.shared.currentToken
    }
    
    func createJSONDecoder() -> JSONDecoder {
        let decoder = JSONDecoder()
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
        decoder.dateDecodingStrategy = .formatted(dateFormatter)
        return decoder
    }
}

// MARK: - Empty Response for DELETE requests
struct EmptyResponse: Decodable {}

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case decodingError
    case unauthorized
    case notFound
    case httpError(Int)
    case serverError(Int)
    case notImplemented
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid server response"
        case .decodingError:
            return "Failed to decode response"
        case .unauthorized:
            return "Authentication required"
        case .notFound:
            return "Resource not found"
        case .httpError(let code):
            return "HTTP error: \(code)"
        case .serverError(let code):
            return "Server error: \(code)"
        case .notImplemented:
            return "This feature is not implemented yet"
        }
    }
}